examples/firewall/firewall.c

```diff
@@ -0,0 +1,377 @@
+        firewall_req->socket_id = rte_socket_id();
+        strcpy(firewall_req->name, name);
+
+        status = onvm_nflib_request_lpm(firewall_req);
+
+        if (status < 0) {
+                rte_exit(EXIT_FAILURE, "Cannot get lpm region for firewall\n");
+        }
+
+        lpm_tbl = rte_lpm_find_existing(name);
+
+        if (lpm_tbl == NULL) {
+                printf("No existing LPM_TBL\n");
+        }
+
+        for (i = 0; i < num_rules; ++i) {
+                ip = rules[i]->src_ip;
+                printf("RULE %d: { ip: %u.%u.%u.%u, depth: %d, action: %d }\n", i, (ip >> 24) & 0xFF, (ip >> 16) & 0xFF, (ip >> 8) & 0xFF, ip & 0xFF, rules[i]->depth,
+                       rules[i]->action);
+                printf("In other words %u\n", rules[i]->src_ip);
+                ret = rte_lpm_add(lpm_tbl, rules[i]->src_ip, rules[i]->depth, rules[i]->action);
+                if (ret < 0) {
+                        printf("ERROR ADDING RULE %d\n", ret);
+                        return 1;
+                }
+        }
+        rte_free(firewall_req);
+
+
+        return 0;
+}
+
+static void lpm_teardown(struct onvm_fw_rule **rules, int num_rules) {
+        int i;
+
+        if (rules) {
+                for (i = 0; i < num_rules; ++i) {
+                        if (rules[i]) free(rules[i]);
+                }
+                free(rules);
+        }
+
+        if (lpm_tbl) {
+                rte_lpm_free(lpm_tbl);
+        }
+}
+
+struct onvm_fw_rule **setup_rules(int *total_rules, char *rules_file) {
+        int ip[4];
+        int num_rules, ret;
+        int i = 0;
+        struct onvm_fw_rule **rules;
+
+        cJSON *rules_json = onvm_config_parse_file(rules_file);
+        cJSON *rules_ip = NULL;
+        cJSON *depth = NULL;
+        cJSON *action = NULL;
+
+        if (rules_json == NULL) {
+                rte_exit(EXIT_FAILURE, "%s file could not be parsed/not found. Assure rules file"
+                                       " the directory to the rules file is being specified.\n", rules_file);
+        }
+
+        num_rules = onvm_config_get_item_count(rules_json);
+        *total_rules = num_rules;
+        rules = (struct onvm_fw_rule **) malloc(num_rules * sizeof(struct onvm_fw_rule *));
+        rules_json = rules_json->child;
+
+        while (rules_json != NULL) {
+                rules_ip = cJSON_GetObjectItem(rules_json, "ip");
+                depth = cJSON_GetObjectItem(rules_json, "depth");
+                action = cJSON_GetObjectItem(rules_json, "action");
+
+                if (rules_ip == NULL) rte_exit(EXIT_FAILURE, "IP not found/invalid\n");
+                if (depth == NULL) rte_exit(EXIT_FAILURE, "Depth not found/invalid\n");
+                if (action == NULL) rte_exit(EXIT_FAILURE, "Action not found/invalid\n");
+
+                rules[i] = (struct onvm_fw_rule *) malloc(sizeof(struct onvm_fw_rule));
+                ret = sscanf(rules_ip->valuestring, "%u.%u.%u.%u", &ip[3], &ip[2], &ip[1], &ip[0]);
```

Comment:
This works but its just a bit too hacky. We need to either change the ip_parse function or make the firewall work using it

Reply:
if we convert from be to CPU on the parsed IP, as well as on the incoming packet, this works without having to change the function